DOI Content Negotiations.js

```diff
@@ -0,0 +1,194 @@
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+function detectSearch(items) {
+	return (filterQuery(items).length > 0);
+}
+
+// return an array of DOIs from the query (items or text)
+function filterQuery(items) {
+	if(!items) return [];
+
+	if(typeof items == 'string' || !items.length) items = [items];
+
+	//filter out invalid queries
+	var dois = [], doi;
+	for(var i=0, n=items.length; i<n; i++) {
+		if(items[i].DOI && (doi = ZU.cleanDOI(items[i].DOI)) ) {
+			dois.push(doi);
+		} else if(typeof items[i] == 'string' && (doi = ZU.cleanDOI(items[i])) ) {
+			dois.push(doi);
+		}
+	}
+	return dois;
+}
+
+function doSearch(items) {
+	var dois = filterQuery(items);
+	if(!dois.length) return;
+	processDOIs(dois);
+}
+
+function processDOIs(dois) {
+	var doi = dois.pop();
+	// by content negotiation we asked for datacite or crossref format, or CSL JSON
+	ZU.doGet('https://doi.org/' + encodeURIComponent(doi), function(text) {
+		if(!text) {
+			return;
+		}
+		Z.debug(text)
+		
+		var trans = Zotero.loadTranslator('import');
+		trans.setString(text);
+		if (text.includes("<crossref")) {
+			//Crossref Unixref
+			trans.setTranslator('93514073-b541-4e02-9180-c36d2f3bb401');
+			trans.setHandler('itemDone', function(obj, item) {
+				item.libraryCatalog = "DOI.org (Crossref)";
+				item.complete();
+			});
+			trans.translate();
+		}
+		else if (text.includes("http://datacite.org/schema")) {
+			//Datacite JSON
+			trans.setTranslator('b5b5808b-1c61-473d-9a02-e1f5ba7b8eef');
+			trans.setHandler('itemDone', function(obj, item) {
+				item.libraryCatalog = "DOI.org (Datacite)";
+				item.complete();
+			});
+			trans.translate();
+		}
+		else {
+			// use CSL JSON translator
+			trans.setTranslator('bc03b4fe-436d-4a1f-ba59-de4d2d7a63f7');
+			trans.setHandler('itemDone', function(obj, item) {
+				item.libraryCatalog = "DOI.org (CSL JSON)";
+				//check if there are potential issues with character encoding and try to fix it
+				//e.g. 10.1057/9780230391116.0016 (en dash in title is presented as escaped unicode)
+				for(var field in item) {
+					if(typeof item[field] != 'string') continue;
+					//check for control characters that should never be in strings from CrossRef
+					if(/[\u007F-\u009F]/.test(item[field])) {
+						item[field] = decodeURIComponent(escape(item[field]));
```

Comment:
Can we do the same here without the [`escape` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape) which one shouldn't use anymore?

Reply:
Not easily. Neither `JSON.parse` ([SO's top answer](https://stackoverflow.com/posts/12869914/revisions)) nor `encodeURIComponent` (which is _almost_ the same as `escape`), work exactly the same way (and don't work for the case referenced here), so we'd have to build a function for this specific purpose. Since `escape` is going nowhere -- it's technically not even deprecated --  I'd be inclined to just leave it.

Reply:
I played a little around and it seems that the following code might work:
```js
var escaped = item[field].replace(/[^0-9A-Za-z ]/g, function(c) {
	return "%" + c.charCodeAt(0).toString(16);
});
item[field] = decodeURIComponent(escaped);
```

However, I don't understand what we do here conceptually. Escaping and then decoding again should more or less give the same string as we have in the beginning...